src/Java.Interop/Java.Interop/JniEnvironment.Types.cs

```diff
@@ -167,6 +170,17 @@ public static void RegisterNatives (JniObjectReference type, JniNativeMethodRegi
 
 			public static void RegisterNatives (JniObjectReference type, JniNativeMethodRegistration [] methods, int numMethods)
 			{
+#if DEBUG && NETCOREAPP
+				foreach (var m in methods) {
+					if (m.Marshaler.GetType ().GenericTypeArguments.Length != 0) {
+						var method  = m.Marshaler.Method;
+						Debug.WriteLine ($"JNIEnv::RegisterNatives() given a generic delegate type.  .NET Core doesn't like this.");
```

Comment:
What is the issue here, do we get some exception during runtime?

Reply:
Yes. :-)

.NET Core doesn't support marshaling generic delegates via P/Invoke.

For example, apply this diff:

```diff
diff --git a/src/Java.Interop/Java.Interop/ManagedPeer.cs b/src/Java.Interop/Java.Interop/ManagedPeer.cs
index 5c0e033c..b6d7aafd 100644
--- a/src/Java.Interop/Java.Interop/ManagedPeer.cs
+++ b/src/Java.Interop/Java.Interop/ManagedPeer.cs
@@ -25,7 +25,7 @@ namespace Java.Interop {
                                        new JniNativeMethodRegistration (
                                                "construct",
                                                ConstructSignature,
-                                               (ConstructMarshalMethod) Construct),
+                                               (Action<IntPtr, IntPtr, IntPtr, IntPtr, IntPtr, IntPtr>) Construct),
                                        new JniNativeMethodRegistration (
                                                "registerNativeMembers",
                                                RegisterNativeMembersSignature,
```

Then run under .NET Core:

```
% dotnet test bin/TestDebug-netcoreapp3.1/Java.Interop-Tests.dll
```

then watch the fireworks:

```
SetUp failed for test fixture Java.InteropTests.JniValueMarshaler_String_ContractTests
System.TypeInitializationException : The type initializer for 'Java.InteropTests.JavaVMFixture' threw an exception.
  ----> System.TypeInitializationException : The type initializer for 'Java.Interop.ManagedPeer' threw an exception.
  ----> System.ArgumentException : The specified Type must not be a generic type definition. (Parameter 'delegate')
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.ConstructorInfo.Invoke(Object[] parameters)
   at NUnit.Framework.Internal.Reflect.Construct(Type type, Object[] arguments)
   at NUnit.Framework.Internal.TypeWrapper.Construct(Object[] args)
   at NUnit.Framework.Internal.Commands.ConstructFixtureCommand.<.ctor>b__0_0(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.BeforeTestCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformOneTimeSetUp()
--TypeInitializationException
   at Java.Interop.ManagedPeer.Init() in …/Java.Interop/src/Java.Interop/Java.Interop/ManagedPeer.cs:line 41
   at Java.Interop.JniRuntime..ctor(CreationOptions options) in …/Java.Interop/src/Java.Interop/Java.Interop/JniRuntime.cs:line 251
   at Java.Interop.JreRuntime..ctor(JreRuntimeOptions builder) in …/Java.Interop/src/Java.Runtime.Environment/Java.Interop/JreRuntime.cs:line 141
   at Java.InteropTests.TestJVM..ctor(String[] jars, Dictionary`2 typeMappings) in …/Java.Interop/tests/TestJVM/TestJVM.cs:line 61
   at Java.InteropTests.JavaVMFixture.CreateJavaVM() in …/Java.Interop/tests/Java.Interop-Tests/Java.Interop/JavaVMFixture.cs:line 15
   at Java.InteropTests.JavaVMFixture..cctor() in …/Java.Interop/tests/Java.Interop-Tests/Java.Interop/JavaVMFixture.Partial.cs:line 12
--ArgumentException
   at System.StubHelpers.MngdNativeArrayMarshaler.ConvertContentsToNative(IntPtr pMarshalState, Object& pManagedHome, IntPtr pNativeHome)
   at Java.Interop.NativeMethods.java_interop_jnienv_register_natives(IntPtr jnienv, IntPtr& thrown, IntPtr type, JniNativeMethodRegistration[] methods, Int32 numMethods)
   at Java.Interop.JniEnvironment.Types._RegisterNatives(JniObjectReference type, JniNativeMethodRegistration[] methods, Int32 numMethods) in …/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.g.cs:line 13148
   at Java.Interop.JniEnvironment.Types.RegisterNatives(JniObjectReference type, JniNativeMethodRegistration[] methods, Int32 numMethods) in …/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.Types.cs:line 188
   at Java.Interop.JniType.RegisterNativeMethods(JniNativeMethodRegistration[] methods) in …/Java.Interop/src/Java.Interop/Java.Interop/JniType.cs:line 151
   at Java.Interop.ManagedPeer..cctor() in …/Java.Interop/src/Java.Interop/Java.Interop/ManagedPeer.cs:line 24
SetUp failed for test fixture Java.InteropTests.MethodBindingTests
System.TypeInitializationException : The type initializer for 'Java.InteropTests.JavaVMFixture' threw an exception.
  ----> System.TypeInitializationException : The type initializer for 'Java.Interop.ManagedPeer' threw an exception.
  ----> System.ArgumentException : The specified Type must not be a generic type definition. (Parameter 'delegate')
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.ConstructorInfo.Invoke(Object[] parameters)
   at NUnit.Framework.Internal.Reflect.Construct(Type type, Object[] arguments)
   at NUnit.Framework.Internal.TypeWrapper.Construct(Object[] args)
   at NUnit.Framework.Internal.Commands.ConstructFixtureCommand.<.ctor>b__0_0(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.BeforeTestCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformOneTimeSetUp()
--TypeInitializationException
   at Java.Interop.ManagedPeer.Init() in …/Java.Interop/src/Java.Interop/Java.Interop/ManagedPeer.cs:line 41
   at Java.Interop.JniRuntime..ctor(CreationOptions options) in …/Java.Interop/src/Java.Interop/Java.Interop/JniRuntime.cs:line 251
   at Java.Interop.JreRuntime..ctor(JreRuntimeOptions builder) in …/Java.Interop/src/Java.Runtime.Environment/Java.Interop/JreRuntime.cs:line 141
   at Java.InteropTests.TestJVM..ctor(String[] jars, Dictionary`2 typeMappings) in …/Java.Interop/tests/TestJVM/TestJVM.cs:line 61
   at Java.InteropTests.JavaVMFixture.CreateJavaVM() in …/Java.Interop/tests/Java.Interop-Tests/Java.Interop/JavaVMFixture.cs:line 15
   at Java.InteropTests.JavaVMFixture..cctor() in …/Java.Interop/tests/Java.Interop-Tests/Java.Interop/JavaVMFixture.Partial.cs:line 12
--ArgumentException
   at System.StubHelpers.MngdNativeArrayMarshaler.ConvertContentsToNative(IntPtr pMarshalState, Object& pManagedHome, IntPtr pNativeHome)
   at Java.Interop.NativeMethods.java_interop_jnienv_register_natives(IntPtr jnienv, IntPtr& thrown, IntPtr type, JniNativeMethodRegistration[] methods, Int32 numMethods)
   at Java.Interop.JniEnvironment.Types._RegisterNatives(JniObjectReference type, JniNativeMethodRegistration[] methods, Int32 numMethods) in …/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.g.cs:line 13148
   at Java.Interop.JniEnvironment.Types.RegisterNatives(JniObjectReference type, JniNativeMethodRegistration[] methods, Int32 numMethods) in …/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.Types.cs:line 188
   at Java.Interop.JniType.RegisterNativeMethods(JniNativeMethodRegistration[] methods) in …/Java.Interop/src/Java.Interop/Java.Interop/JniType.cs:line 151
   at Java.Interop.ManagedPeer..cctor() in …/Java.Interop/src/Java.Interop/Java.Interop/ManagedPeer.cs:line 24
```

`.MngdNativeArrayMarshaler.ConvertContentsToNative()` throws an ArgumentException if you attempt to marshal a generic delegate type.

This is also why `Java.Interop.Export.dll` is broken under .NET Core.